### PR TITLE
refactor: checkpoints

### PR DESCRIPTION
**Pull request by lukealvoeiro**
This PR introduces a large rewrite of checkpoints, ensuring they are always in sync with messages on the exchange.

#### Why?

As the exchange grows in side, it is increasingly important to trim its beginning (as these messages are less relevant). We also need to ensure the exchange has an accurate representation of the tokens it's used up so far, so we can avoid:
1. Sending the LLM too many tokens
2. Sending the LLM requests that are over a certain size (altering the speed of its response)

This PR makes managing checkpoints and messages much easier, introducing the following methods on the exchange:
- `.pop_first_message()`
- `.pop_last_message()`
- `.pop_first_checkpoint()`
- `.pop_last_checkpoint()`

I have also included a large number of tests that cover the different ways that messages and checkpoints could fall out of sync. Now that these are covered, we hope to not run into an regressions in this behavior.

#### Note to reviewers

There are a couple of sections marked with `TODO` comments. These have been intentionally left behind as they are not essential features in order to be backwards compatible. We can instead focus on implementing these over the coming weeks.

### CODE REVIEW

File: src/exchange/moderators/summarizer.py [hunk 1]
```diff
@@ -1,135 +1,49 @@
...
+        num_messages_to_remove = len(messages_to_summarize)
+
+        # the llm will throw an error if the last message isn't a user message
+        if messages_to_summarize[-1].role == "assistant" and (not messages_to_summarize[-1].tool_use):
+            messages_to_summarize.append(Message.user("Summarize our the above conversation"))
+
+        summarizer_exchange = exchange.replace(
+            system=Message.load("summarizer.jinja").text,
+            moderator=PassiveModerator(),  # note-pr-reviewer: this was previously ContextTruncate
+            model=self.model,
```
Comment (lukealvoeiro) [inline src/exchange/moderators/summarizer.py:28]: note: this used to be a truncate moderator, but we now are using a passive one. why? because we know the message here will always be under the token limit

File: src/exchange/checkpoint.py [hunk 2]
```diff
@@ -1,11 +1,67 @@
...
+        return Checkpoint(start_index=self.start_index, end_index=self.end_index, token_count=self.token_count)
+
+
+@define
+class CheckpointData:
+    """Aggregates all information about checkpoints"""
+
+    total_token_count: int = field(default=0)
+    checkpoints: List[Checkpoint] = field(default=[])
+    message_index_offset: int = field(default=0)
```
Comment (zakiali) [inline src/exchange/checkpoint.py:33]: lets add a description of this (and other attributes, but those are more self explanatory)

File: src/exchange/checkpoint.py [hunk 3]
```diff
@@ -1,11 +1,67 @@
...
+
+    def __copy__(self) -> "CheckpointData":
+        """Returns a deep copy of the CheckpointData object."""
+        return CheckpointData(
+            total_token_count=self.total_token_count,
+            checkpoints=deepcopy(self.checkpoints),
```
Comment (zakiali) [inline src/exchange/checkpoint.py:39]: for expectations on behavior, maybe changing this to be a copy() instead of deepcopy() or even a copy(self)

File: src/exchange/exchange.py [hunk 4]
```diff
@@ -38,10 +39,10 @@ class Exchange:
     provider: Provider
     model: str
     system: str
-    moderator: Moderator = field(default=ContextTruncate())
+    moderator: Moderator = field(default=PassiveModerator())
```
Comment (zakiali) [inline src/exchange/exchange.py:42]: any reason for changing this to Passive? Truncate is probably a better user experience

File: src/exchange/exchange.py [hunk 5]
```diff
@@ -51,8 +52,10 @@ def replace(self, **kwargs: Dict[str, Any]) -> "Exchange":
         """Make a copy of the exchange, replacing any passed arguments"""
         if kwargs.get("messages") is None:
             kwargs["messages"] = deepcopy(self.messages)
-        if kwargs.get("checkpoints") is None:
-            kwargs["checkpoints"] = deepcopy(self.checkpoints)
+        if kwargs.get("checkpoint_data") is None:
```
Comment (zakiali) [inline src/exchange/exchange.py:55]: This breaks backwards compatibility, specifically breaking some uses in Goose: https://github.com/search?q=repo%3Asquare%2Fgoose%20checkpoints%3D&type=code. We'll need to update those references and anywhere else they might be (custom plugins repo's).

File: src/exchange/exchange.py [hunk 6]
```diff
@@ -158,16 +167,144 @@ def add_tool_use(self, tool_use: ToolUse) -> None:
...
+        if len(self.messages) - new_start_index > 1:
+            # this case may come up if the user manipulates the message queue. e.g. if the
+            # user performs the following actions:
+            # 1. adds a message
+            # 2. calls pop_last_checkpoint
+            # 3. adds another message
+            # 4. calls reply
+            self.checkpoint_data.checkpoints.append(
+                Checkpoint(
+                    start_index=new_start_index + self.checkpoint_data.message_index_offset,
```
Comment (zakiali) [inline src/exchange/exchange.py:202]: Maybe a comment about the -2? If i'm thinking about this correctly, we're adding in the first checkpoint (so first group of messages). The last message (corresponding to output_tokens) would be -1.

File: src/exchange/exchange.py [hunk 7]
```diff
@@ -158,16 +167,144 @@ def add_tool_use(self, tool_use: ToolUse) -> None:
...
+            raise ValueError("There are no messages to pop")
+        if len(self.checkpoint_data.checkpoints) == 0:
+            raise ValueError("There must be at least one checkpoint to pop the first message")
+
+        # get the start and end indexes of the first checkpoint, use these to remove message
+        first_checkpoint = self.checkpoint_data.checkpoints[0]
+        first_checkpoint_start_index = first_checkpoint.start_index - self.checkpoint_data.message_index_offset
+
+        # check if the first message is part of the first checkpoint
+        if first_checkpoint_start_index == 0:
```
Comment (zakiali) [inline src/exchange/exchange.py:251]: Is it so that if this is not the case, we've goofed somewhere and are not in a good state? Is there a way we could get here (I don't think so), but maybe we add a check for that to catch it

File: src/exchange/exchange.py [hunk 8]
```diff
@@ -158,16 +167,144 @@ def add_tool_use(self, tool_use: ToolUse) -> None:
...
+            self.checkpoint_data.message_index_offset = 0
+
+    def pop_last_checkpoint(self) -> Tuple[Checkpoint, List[Message]]:
+        """
+        Reverts the exchange back to the last checkpoint, removing associated messages
+        """
+        removed_checkpoint = self.checkpoint_data.checkpoints.pop()
+        # pop messages until we reach the start of the next checkpoint
+        while len(self.messages) > removed_checkpoint.start_index - self.checkpoint_data.message_index_offset:
+            self.messages.pop()
```
Comment (zakiali) [inline src/exchange/exchange.py:269]: missing return values

File: src/exchange/exchange.py [hunk 9]
```diff
@@ -158,16 +167,144 @@ def add_tool_use(self, tool_use: ToolUse) -> None:
...
+            Checkpoint(
+                start_index=new_index,
+                end_index=new_index,
+                token_count=token_count,
+            ),
+        )
+        self.checkpoint_data.message_index_offset = new_index
+
+    @property
+    def is_allowed_to_call_llm(self) -> bool:
```
Comment (zakiali) [inline src/exchange/exchange.py:306]: This is kinda LLM dependent? You can call gpt4o without any messages for example

File: src/exchange/moderators/summarizer.py [hunk 10]
```diff
@@ -1,135 +1,52 @@
...
-from exchange.checkpoint import Checkpoint
-from exchange.moderators import ContextTruncate, Moderator, PassiveModerator
+from exchange.moderators import ContextTruncate, PassiveModerator
 
-MAX_TOKENS = 112000
-SUMMARIZATION_OFFSET = 40000  # Keep a max of this many tokens
+# currently this is the point at which we start to summarize, so
+# so once we get to this token size the token count will exceed this
+# by a little bit
+MAX_TOKENS = 70000
```
Comment (zakiali) [inline src/exchange/moderators/summarizer.py:9]: Same comment as for Truncate. We should start summarizing at a higher token count, but also having a knob to summarize a fraction (or offset) of the messages

File: src/exchange/moderators/summarizer.py [hunk 11]
```diff
@@ -1,135 +1,52 @@
...
 
 
-def pop_checkpoint(
```
Comment (zakiali) [inline src/exchange/moderators/summarizer.py:11]: Nice that this gets cleaned up!

File: src/exchange/moderators/summarizer.py [hunk 12]
```diff
@@ -1,135 +1,52 @@
...
+        if messages_to_summarize[-1].role == "assistant" and (not messages_to_summarize[-1].tool_use):
+            messages_to_summarize.append(Message.user("Summarize our the above conversation"))
+
+        summarizer_exchange = exchange.replace(
+            system=Message.load("summarizer.jinja").text,
+            moderator=PassiveModerator(),
+            model=self.model,
+            messages=messages_to_summarize,
+            # checkpoint_data=CheckpointData(),
+            # TODO: figure out why the summarizer exchange has checkpoint data that is not empty
```
Comment (zakiali) [inline src/exchange/moderators/summarizer.py:34]: What is it set too? Is this because the messages_to_summarize is a subset of the messages? Haven't looked deeply at this

File: src/exchange/moderators/truncate.py [hunk 13]
```diff
@@ -1,59 +1,70 @@
...
-        # there is still one entry in the first checkpoint only
-        checkpoint.token_count = checkpoint.latest_generated_tokens
-        checkpoint.end_index = exclude_last
-        exchange.checkpoints.insert(0, checkpoint)
-
-    return exchange
+# currently this is the point at which we start to truncate, so
+# so once we get to this token size the token count will exceed this
+# by a little bit
+MAX_TOKENS = 70000
```
Comment (zakiali) [inline src/exchange/moderators/truncate.py:11]: This should be larger. 128K is the standard context lengths, and we have a max limit on tool usage at 16000 (https://github.com/square/exchange/blob/ed3a6ab8ff4d414494980da4cfc3604264ec3e9b/src/exchange/exchange.py#L20), which is where we got the 112000 before. We'll start loosing too much context if we start at 70K.

File: src/exchange/moderators/truncate.py [hunk 14]
```diff
@@ -1,59 +1,70 @@
...
+            self.last_system_prompt = exchange.system
+
+        if not self.system_prompt_token_count or is_different_system_prompt:
             # calculate the system prompt tokens (includes functions etc...)
             _system_token_exchange = exchange.replace(
                 messages=[],
-                checkpoints=[],
+                checkpoint_data=CheckpointData(),
                 moderator=PassiveModerator(),
                 model=self.model if self.model else exchange.model,
```
Comment (zakiali) [inline src/exchange/moderators/truncate.py:44]: This should prob be the same model as the initial exchange so that token content is consistent with what we're expecting. Between models on openai, this is ok, but if we use claude where while the original is gpt-4, there will be discrepancies (don't expect to be major though)

File: src/exchange/moderators/truncate.py (hunk 13 above)
Comment (lukealvoeiro) [reply src/exchange/moderators/truncate.py:11]: Awesome, thanks for explaining that! I'll drop some comments in the code for posterity.

File: src/exchange/checkpoint.py (hunk 3 above)
Comment (lukealvoeiro) [reply src/exchange/checkpoint.py:39]: actually we don't use the __copy__ methods at all - removing them to avoid confusion.

File: src/exchange/exchange.py (hunk 4 above)
Comment (lukealvoeiro) [reply src/exchange/exchange.py:42]: Ah didn't catch this in my proof-read - meant to change it back. Though thoughts on making it `ContextSummarize()`?

File: src/exchange/exchange.py (hunk 6 above)
Comment (lukealvoeiro) [reply src/exchange/exchange.py:202]: yup, just added some more detail here. good callout on the explanation.

File: src/exchange/exchange.py (hunk 7 above)
Comment (lukealvoeiro) [reply src/exchange/exchange.py:251]: We can totally reach both `first_checkpoint_start_index == 0` and `first_checkpoint_start_index != 0`. Just depicted an example below:

```
note: [] denotes list of messages, () denotes how checkpoints are grouped

if we start off with:
[(1,2,3),(4),(5)]

pop_first_message()
> first_checkpoint_start_index == 0
> and the resulting array: [2,3,(4),(5)]

pop_first_message() again
> first_checkpoint_start_index == 2
> and the resulting array: [3,(4),(5)]
```

File: src/exchange/moderators/summarizer.py (hunk 12 above)
Comment (lukealvoeiro) [reply src/exchange/moderators/summarizer.py:34]: The idea here was to create a brand new instance of `CheckpointData`. This is helpful because when we then call `.reply()` to get the summary, we have **two** checkpoints: `(0..n-1) & (n)` - where n is the summary message. Just implemented the `TODO` captured in this codeblock in the most recent change to this PR.

File: src/exchange/exchange.py (hunk 5 above)
Comment (lukealvoeiro) [reply src/exchange/exchange.py:55]: Good catch. Will update once this PR gets merged in. We'll a major version bump to indicate the breaking changes, even though they are minor breaking changes.

File: src/exchange/moderators/truncate.py (hunk 14 above)
Comment (lukealvoeiro) [reply src/exchange/moderators/truncate.py:44]: Good call, I just added a check for this on the `.replace()` method - any calls to `.replace()` that specify `model` in the `kwargs` should also specify `checkpoint_data`.

I'm not convinced that we should override to usingt he exchange's model here, as it defeats the purpose of specifying an accelerator model in the profile being used. Instead, we expect folks will configure profiles in a meaningful way, and our scripting should add enough token count padding to account for some user errors.

**Review by codefromthecrypt — COMMENTED** [review]
drive-by: I might add to the integration test an edge case, so that we don't lose the system prompt or other things that could interfere with tool calls. This is in addition to the unit tests you have here, as it helps us know end-to-end works.

File: src/exchange/exchange.py (hunk 9 above)
Comment (lukealvoeiro) [reply src/exchange/exchange.py:306]: Hmmm, maybe this belongs in the Provider class 🤔. I'm going to comment out it's usage now and mark it with a `TODO` so we can revisit it at a later point.

File: src/exchange/moderators/summarizer.py (hunk 10 above)
Comment (michaelneale) [reply src/exchange/moderators/summarizer.py]: should ie also be per provider? claude has a much higher context window for example

File: src/exchange/exchange.py (hunk 4 above)
Comment (zakiali) [reply src/exchange/exchange.py:43]: Yea eventually. I think we still need feedback on it to see how well it's working for people

File: src/exchange/moderators/truncate.py (hunk 14 above)
Comment (zakiali) [reply src/exchange/moderators/truncate.py:44]: I just meant for this specific case to calculate the system message token count, but I guess this effects everything. +1 on users using self consistent profiles

File: src/exchange/moderators/summarizer.py (hunk 10 above)
Comment (zakiali) [reply src/exchange/moderators/summarizer.py]: yea, definitely per provider. We need a way to set these in the profiles (on a backlog) so users can set these values directly instead of using the defaults.

File: src/exchange/moderators/summarizer.py (hunk 12 above)
Comment (zakiali) [reply src/exchange/moderators/summarizer.py]: makes sense. Was this pushed up? I don't see the changes described.

**Review by zakiali — APPROVED** [review]

File: src/exchange/moderators/summarizer.py (hunk 12 above)
Comment (lukealvoeiro) [reply src/exchange/moderators/summarizer.py]: Yup, changes are pushed up. You should see it's changed from 
```
...
            model=self.model,
            messages=messages_to_summarize,
            # checkpoint_data=CheckpointData(),
            # TODO: figure out why the summarizer exchange has checkpoint data that is not empty
```
to 
```
...
            model=self.model,
            messages=messages_to_summarize,
            checkpoint_data=CheckpointData(),
```

File: src/exchange/moderators/summarizer.py (hunk 10 above)
Comment (lukealvoeiro) [reply src/exchange/moderators/summarizer.py]: Agreed, we need to add two new things for providers: 
- `max_tokens` (variable)
- `is_allowed_to_call_llm` (method)

I've annotated both of these areas in the code with TODOs

File: src/exchange/exchange.py (hunk 4 above)
Comment (lukealvoeiro) [reply src/exchange/exchange.py:43]: Ok, just set it back to `ContextTruncate` and we can revisit later.

File: src/exchange/moderators/truncate.py (hunk 14 above)
Comment (lukealvoeiro) [reply src/exchange/moderators/truncate.py:45]: > Good call, I just added a check for this on the .replace() method - any calls to .replace() that specify model in the kwargs should also specify checkpoint_data.

I actually ended up removing this, as it causes too many new errors in Goose. This is indicative of us already falling into patterns of mutating the Exchange in potentially unsafe ways.

File: src/exchange/checkpoint.py [hunk 15]
```diff
@@ -1,11 +1,63 @@
...
+
+    def __deepcopy__(self, _) -> "Checkpoint":
+        """
+        Returns a deep copy of the Checkpoint object.
+        """
+        return Checkpoint(start_index=self.start_index, end_index=self.end_index, token_count=self.token_count)
+
+
+@define
+class CheckpointData:
```
Comment (codefromthecrypt) [inline src/exchange/checkpoint.py:22]: completely style, but as this is an aggregate, words that imply plural can help vs something like Data. At first, esp as there is a function about "message" in here, I thought there may be data, like messages in here..

```suggestion
class Checkpoints:
```

File: src/exchange/checkpoint.py [hunk 16]
```diff
@@ -1,11 +1,63 @@
...
+
+    # in order list of individual checkpoints in the exchange
+    checkpoints: List[Checkpoint] = field(factory=list)
+
+    # the offset to apply to the message index when calculating the last message index
+    # this is useful because messages on the exchange behave like a queue, where you can only
+    # pop from the left or right sides. This offset allows us to map the checkpoint indices
+    # to the correct message index, even if we have popped messages from the left side of
+    # the exchange in the past. we reset this offset to 0 when we empty the checkpoint data.
+    message_index_offset: int = field(default=0)
```
Comment (codefromthecrypt) [inline src/exchange/checkpoint.py:37]: I'm curious why this isn't on the exchange instead. If it were, then I feel this object is more tidy, an aggregation of checkpoints... Also, it would be a lot easier to understand its relationship to messages (as they are defined on the exchange). Food for thought and I'm sure I'm missing something.

**Review by codefromthecrypt — APPROVED** [review]
just a last minute bikeshed 😎

File: src/exchange/checkpoint.py (hunk 16 above)
Comment (lukealvoeiro) [reply src/exchange/checkpoint.py:37]: I totally agree! The `Exchange` class should be a clean representation of messages and checkpoints that stay in sync within it's same instance. However, our current implementation of the `Exchange` class is a frozen one - this means that variables such as  `message_index_offset` and `total_token_count` cannot be updated without hacky workarounds. Will be having a discussion with the team to hopefully change this.

File: src/exchange/checkpoint.py (hunk 15 above)
Comment (lukealvoeiro) [reply src/exchange/checkpoint.py:22]: Good idea!
